<a href="https://colab.research.google.com/github/KeshriPooja/PythonCode/blob/master/CoronaVirus_data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Corona Virus Dashboard
 In 2020 There is a virus which spread all over the world. C oviod-19

[More info](https://www.covid19india.org/)

## Data
Data source is collected from the following [API](https://api.covid19india.org/csv/) 

In [0]:
import pandas as pd
import numpy as np

In [0]:
raw_data=pd.read_csv("https://api.covid19india.org/csv/latest/raw_data.csv")

In [3]:
raw_data.shape

(28183, 21)

In [4]:
raw_data.describe()

Patient Number  Estimated Onset Date     Num cases
count    27891.000000                   0.0  28183.000000
mean     13946.000000                   NaN      0.989639
std       8051.582515                   NaN      8.697576
min          1.000000                   NaN   -602.000000
25%       6973.500000                   NaN      1.000000
50%      13946.000000                   NaN      1.000000
75%      20918.500000                   NaN      1.000000
max      27891.000000                   NaN    602.000000

In [5]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28183 entries, 0 to 28182
Data columns (total 21 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Patient Number                             27891 non-null  float64
 1   State Patient Number                       2997 non-null   object 
 2   Date Announced                             28183 non-null  object 
 3   Estimated Onset Date                       0 non-null      float64
 4   Age Bracket                                2346 non-null   object 
 5   Gender                                     5313 non-null   object 
 6   Detected City                              2350 non-null   object 
 7   Detected District                          21198 non-null  object 
 8   Detected State                             28182 non-null  object 
 9   State code                                 28182 non-null  object 
 10  Current Status        

In [6]:
raw_data.columns

Index(['Patient Number', 'State Patient Number', 'Date Announced',
       'Estimated Onset Date', 'Age Bracket', 'Gender', 'Detected City',
       'Detected District', 'Detected State', 'State code', 'Current Status',
       'Notes', 'Contracted from which Patient (Suspected)', 'Nationality',
       'Type of transmission', 'Status Change Date', 'Source_1', 'Source_2',
       'Source_3', 'Backup Notes', 'Num cases'],
      dtype='object')

In [7]:
raw_data.tail()

Patient Number State Patient Number  ... Backup Notes  Num cases
28178             NaN                  NaN  ...          NaN         11
28179             NaN                  NaN  ...          NaN          1
28180             NaN                  NaN  ...          NaN         -1
28181             NaN                  NaN  ...          NaN          6
28182             NaN                  NaN  ...          NaN         -6

[5 rows x 21 columns]

### Details of Intial Paitent founded coronavirus 

First Case of corona in india is on 2020-01-30 00:00:00  found in Thrissur in Kerala is F with age of 20. The case to be known as Student from Wuhan

In [8]:
'''First Patient of corona virus is found on {} in {} in {} is {} with age of {}. The case to be known as {}'''.format(raw_data.iloc[0]['Date Announced'],raw_data.iloc[0]['Detected District'],raw_data.iloc[0]['Detected State'],raw_data.iloc[0]['Gender'],raw_data.iloc[0]['Age Bracket'],raw_data.iloc[0]['Backup Notes'])

'First Patient of corona virus is found on 30/01/2020 in Thrissur in Kerala is F with age of 20. The case to be known as Student from Wuhan'

District wise list of the cases  from From 2020-01-30 to 2020-03-14.

In [9]:
''' From {} to {}'''.format(raw_data.iloc[0]['Date Announced'],raw_data.iloc[99]['Date Announced'])

' From 30/01/2020 to 14/03/2020'

In [0]:
#Converting Date Announced value to python datetime object by which datetime operation will be perform.
raw_data['Date Announced']=pd.to_datetime(raw_data['Date Announced'], format='%d/%m/%Y')

In [15]:
raw_data.sort_values(['Date Announced']).head(100)

Patient Number  ... Num cases
0              1.0  ...         1
1              2.0  ...         1
2              3.0  ...         1
3              4.0  ...         1
4              5.0  ...         1
..             ...  ...       ...
97            98.0  ...         1
96            97.0  ...         1
95            96.0  ...         1
94            95.0  ...         1
93            94.0  ...         1

[100 rows x 21 columns]

In [18]:
raw_data[raw_data['Detected District']=='Italians']

Patient Number  ... Num cases
5              6.0  ...         1
6              7.0  ...         1
7              8.0  ...         1
8              9.0  ...         1
9             10.0  ...         1
10            11.0  ...         1
11            12.0  ...         1
12            13.0  ...         1
13            14.0  ...         1
14            15.0  ...         1
15            16.0  ...         1
16            17.0  ...         1
17            18.0  ...         1
18            19.0  ...         1
19            20.0  ...         1
20            21.0  ...         1

[16 rows x 21 columns]

In [25]:
raw_data.sort_values(['Date Announced']).head(100).groupby('Detected City')['Patient Number'].count().to_frame().sort_values(['Patient Number'], ascending=False)

Patient Number
Detected City                           
Gurugram                              14
Agra                                   9
Ranni                                  7
Pune                                   7
Bengaluru                              5
Nagpur                                 4
Jaipur                                 4
Pimpri-Chinchwad                       3
Mumbai                                 3
Janakpuri                              3
Kannur                                 2
Thrissur                               2
Ladakh                                 2
Kozhencherry                           2
Kochi                                  2
Yavatmal                               2
Jammu                                  2
Hyderabad                              2
Ghaziabad                              2
Chengalam                              2
Delhi                                  1
Varkala                                1
Uttam Nagar                            1
Tri Nagar                              1
Amritsar                               1
Thiruvananthapuram                     1
Thane - Vashi                          1
Chuchot                                1
Kalyan-Dombivali                       1
Panvel (Kamothe)                       1
Kancheepuram                           1
Nedumangad                             1
Navi Mumbai                            1
East Delhi (Mayur Vihar)               1
Lucknow                                1
Alappuzha                              1
Kasaragod                              1
Kalaburagi                             1
Nellore                                1

From the above table we can say that from the 1st 100 cases reported in india have  highest  number of cases come from Gurugram after that from agra, ranni, pune.
So we can say that Gurugram is responsible for spreading corona virus in 1st 100 day phase.

### Statewise list of corona virus from 1st 100 cases

In [23]:
raw_data.sort_values(['Date Announced']).head(100).groupby('Detected State')['Patient Number'].count().to_frame().sort_values(['Patient Number'], ascending=False)

Patient Number
Detected State                   
Maharashtra                    24
Kerala                         22
Haryana                        14
Uttar Pradesh                  13
Delhi                           7
Karnataka                       6
Rajasthan                       4
Ladakh                          3
Jammu and Kashmir               2
Telangana                       2
Andhra Pradesh                  1
Punjab                          1
Tamil Nadu                      1

From the above table we can say that Maharashtra have highest number of cases in 1st 100 verified cases. There are 24 cases from Maharashtra out of 100 so ~41% people are effected in 1st 100 person from Maharastra in their different cities like pune, Mumbai, Nagpur.

### Cases from Suspeceted Patient

we want to see which Patient are responsible for other paitent.

In [26]:
raw_data.groupby('Contracted from which Patient (Suspected)')['Patient Number'].count().to_frame()

Patient Number
Contracted from which Patient (Suspected)                         
E0                                                             960
E1                                                              11
KA-C0                                                            6
P1007                                                            1
P1013                                                            1
...                                                            ...
https://twitter.com/ArogyaAndhra/status/1249575...               1
https://twitter.com/ArogyaAndhra/status/1249575...               1
https://twitter.com/ArogyaAndhra/status/1249575...               1
https://twitter.com/ArogyaAndhra/status/1249575...               1
https://twitter.com/ArogyaAndhra/status/1249575...               1

[224 rows x 1 columns]

E0 is a patient who is from indonesia infeceted 960 people in india. Below are the sample to whom E0 is infected.

In [27]:
raw_data.query('`Contracted from which Patient (Suspected)`=="E0"').head(10)

Patient Number  ... Num cases
140           141.0  ...         1
162           163.0  ...         1
163           164.0  ...         1
164           165.0  ...         1
165           166.0  ...         1
166           167.0  ...         1
167           168.0  ...         1
168           169.0  ...         1
250           252.0  ...         1
251           253.0  ...         1

[10 rows x 21 columns]

## Analysis of Covid  data before and after the locked down phase

In India from 21st of March locked down started and we are assuming that there will be 14 days time boundary after that a patient will  know he/she has corona virus or not.

In [0]:
statewise_data=pd.read_csv('https://api.covid19india.org/csv/latest/state_wise_daily.csv')

In [43]:
'''Statewise data of each day starting from {} to {}'''.format(statewise_data.iloc[0]['Date'],statewise_data.iloc[-1]['Date'])

'Statewise data of each day starting from 14-Mar-20 to 13-May-20'

In [57]:
statewise_data.head()

Date     Status  TT  AN  AP  AR  AS  BR  ...  RJ  SK  TN  TG  TR  UP  UT  WB
0 2020-03-14  Confirmed  81   0   1   0   0   0  ...   3   0   1   1   0  12   0   0
1 2020-03-14  Recovered   9   0   0   0   0   0  ...   1   0   0   0   0   4   0   0
2 2020-03-14   Deceased   2   0   0   0   0   0  ...   0   0   0   0   0   0   0   0
3 2020-03-15  Confirmed  27   0   0   0   0   0  ...   1   0   0   2   0   1   0   0
4 2020-03-15  Recovered   4   0   0   0   0   0  ...   2   0   0   1   0   0   0   0

[5 rows x 40 columns]

In [0]:
statewise_data['Date']=pd.to_datetime(statewise_data['Date'])

### Melt Function
Melt function is use to convert dataframe. It fix up any columns and associate it with all columns to create a new structured dataframe.



In [0]:

s1=statewise_data.melt(id_vars=['Date','Status'])

In [59]:
s1.head()

Date     Status variable  value
0 2020-03-14  Confirmed       TT   81.0
1 2020-03-14  Recovered       TT    9.0
2 2020-03-14   Deceased       TT    2.0
3 2020-03-15  Confirmed       TT   27.0
4 2020-03-15  Recovered       TT    4.0

## State wise analysis

From the data we see that total number of cases is 78055.

Below the statewise confirmed data are given.

In [74]:
s1.query('Status=="Confirmed"').groupby(['variable']).sum().sort_values('value', ascending=False).iloc[1:,:]

value
variable         
MH        25922.0
GJ         9268.0
TN         9227.0
DL         7998.0
RJ         4328.0
MP         4173.0
UP         3758.0
WB         2290.0
AP         2137.0
PB         1924.0
TG         1367.0
JK          971.0
KA          959.0
BR          953.0
HR          793.0
OR          538.0
KL          535.0
CH          191.0
JH          177.0
TR          154.0
AS           80.0
UT           72.0
HP           68.0
CT           59.0
LA           43.0
AN           33.0
PY           13.0
ML           13.0
GA            7.0
MN            2.0
MZ            1.0
DN            1.0
AR            1.0
NL            0.0
SK            0.0
DD            0.0
LD            0.0

In [0]:
ctab=pd.crosstab(s1['variable'],s1['Status'],s1['value'],aggfunc='sum')

In [0]:
ctab=ctab.reset_index()

In [0]:
ctab['Mortality']=ctab['Deceased']/ctab['Confirmed']

In [88]:
ctab.sort_values(['Mortality'], ascending=False)

Status variable  Confirmed  Deceased  Recovered  Mortality
37           WB     2290.0     207.0      702.0   0.090393
21           ML       13.0       1.0       11.0   0.076923
11           GJ     9268.0     566.0     3562.0   0.061070
23           MP     4173.0     233.0     2004.0   0.055835
12           HP       68.0       3.0       36.0   0.044118
20           MH    25922.0     974.0     5547.0   0.037574
16           KA      959.0      33.0      451.0   0.034411
34           TT    78055.0    2549.0    26418.0   0.032656
29           RJ     4328.0     121.0     2573.0   0.027957
31           TG     1367.0      34.0      939.0   0.024872
35           UP     3758.0      86.0     1965.0   0.022885
1            AP     2137.0      47.0     1142.0   0.021993
14           JH      177.0       3.0       87.0   0.016949
27           PB     1924.0      32.0      200.0   0.016632
5            CH      191.0       3.0       30.0   0.015707
36           UT       72.0       1.0       46.0   0.013889
13           HR      793.0      11.0      417.0   0.013871
8            DL     7998.0     106.0     2858.0   0.013253
3            AS       80.0       1.0       40.0   0.012500
15           JK      971.0      10.0      466.0   0.010299
17           KL      535.0       4.0      490.0   0.007477
4            BR      953.0       7.0      400.0   0.007345
32           TN     9227.0      64.0     2176.0   0.006936
26           OR      538.0       3.0      143.0   0.005576
33           TR      154.0       0.0        2.0   0.000000
0            AN       33.0       0.0       33.0   0.000000
28           PY       13.0       0.0        9.0   0.000000
24           MZ        1.0       0.0        1.0   0.000000
22           MN        2.0       0.0        2.0   0.000000
10           GA        7.0       0.0        7.0   0.000000
9            DN        1.0       0.0        1.0   0.000000
6            CT       59.0       0.0       55.0   0.000000
2            AR        1.0       0.0        1.0   0.000000
18           LA       43.0       0.0       22.0   0.000000
7            DD        0.0       0.0        0.0        NaN
19           LD        0.0       0.0        0.0        NaN
25           NL        0.0       0.0        0.0        NaN
30           SK        0.0       0.0        0.0        NaN

From the above table we can conclude that  west bengal has highest number of mortality rate ~9%.


The plot below represent distribution of confirmed cases of statewise.

In [85]:
import plotly.express as px
fig=px.bar(ctab.query('variable!="TT"'), x='variable', y='Confirmed',hover_data=['Deceased','Mortality'])
fig.show()

In [86]:
fig2=px.bar(ctab.query('variable!="TT"'), x='variable', y='Deceased',hover_data=['Confirmed','Mortality'])
fig2.show()

In [0]:
"statewisetest_data=pd.read_csv('https://api.covid19india.org/csv/latest/statewise_tested_numbers_data.csv')

In [41]:
'''Statewise Test data start from {} to {}'''.format(statewisetest_data.iloc[0]['Updated On'],statewisetest_data.iloc[-1]['Updated On'])

'Statewise Test data start from 17/04/2020 to 13/05/2020'

In [33]:
statewisetest_data.head()

Updated On                        State  ...  Unnamed: 22 Unnamed: 23
0  17/04/2020  Andaman and Nicobar Islands  ...          NaN         NaN
1  24/04/2020  Andaman and Nicobar Islands  ...       1000.0   1000000.0
2  27/04/2020  Andaman and Nicobar Islands  ...          NaN         NaN
3  01/05/2020  Andaman and Nicobar Islands  ...          NaN         NaN
4  02/04/2020               Andhra Pradesh  ...          NaN         NaN

[5 rows x 24 columns]

In [0]:
casetime_seriesdata=pd.read_csv('https://api.covid19india.org/csv/latest/case_time_series.csv')

In [0]:
casetime_seriesdata.head()